### PR TITLE
WIP: Initial work on private dns zone associations

### PR DESCRIPTION
**Pull request by justinsb**
Not yet wired up

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/516)

### CODE REVIEW

**Comment by kris-nova** [comment]
@justinsb - Update on this?

File: upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	VPC     *VPC
+}
+
+var _ fi.CompareWithID = &AutoscalingGroup{}
+
+func (e *DNSZoneAssociation) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *DNSZoneAssociation) Find(c *fi.Context) (*DNSZoneAssociation, error) {
```
Comment (JuJu227) [inline upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go]: I probably just don't understand the code well enough but this seems a tad odd. Are we expecting nil for both the object and the error? Would it make sense to return slice of DNSZoneAssociation  or  have three return variables given that find() can generally have three different states. 
1) Something went wrong
2) Something was found 
3) Nothing was found

File: upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	return nil, nil
+}
+
+func (e *DNSZoneAssociation) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (s *DNSZoneAssociation) CheckChanges(a, e, changes *DNSZoneAssociation) error {
```
Comment (JuJu227) [inline upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go]: From a readability perspective: 
   1) func (e *DNSZoneAssociation) -> func (e *DNSZoneAssociation) should be more consistent with naming convention. 
   2) Single letter argurements/variable names becomes hard to follow.  Only referring to the  CheckChanges(a, e, changes *DNSZoneAssociation) porition of the signature

File: upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+//go:generate fitask -type=DNSZoneAssociation
+type DNSZoneAssociation struct {
+	Name    *string
+	DNSZone *DNSZone
+	VPC     *VPC
+}
+
+var _ fi.CompareWithID = &AutoscalingGroup{}
+
+func (e *DNSZoneAssociation) CompareWithID() *string {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go]: Yes, we're implementing `fi.CompareWithID`

File: upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go (hunk 1 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go]: We either return `found, nil` or `nil, error` or `nil, nil`.  The last case is not found, no error (i.e. does not exist).  I think that covers the 3 cases pretty nicely.

File: upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go (hunk 2 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go]: a = actual, e = expected.  It's what we're dealing with everywhere here, so it becomes a pattern.

**Comment by kris-nova** [comment]
@justinsb Can we get this conflict tidied up please?

**Comment by kris-nova** [comment]
@justinsb ping #2

**Comment by justinsb** [comment]
Part of #1264
